internal/speakerlist/speakerlist.go

```diff
@@ -0,0 +1,198 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package speakerlist
+
+import (
+	"crypto/sha256"
+	"errors"
+	golog "log"
+	"reflect"
+	"strconv"
+	"sync/atomic"
+	"time"
+
+	"go.universe.tf/metallb/internal/k8s"
+	v1 "k8s.io/api/core/v1"
+
+	gokitlog "github.com/go-kit/kit/log"
+	"github.com/hashicorp/memberlist"
+)
+
+// SpeakerList gives you the list of healthy speaker
+type SpeakerList struct {
+	l           gokitlog.Logger
+	client      *k8s.Client
+	mList       *memberlist.Memberlist
+	k8sSpeakers atomic.Value
+	eventCh     chan memberlist.NodeEvent
+	stopCh      chan struct{}
+}
+
+// New creates a new SpeakerList
+func New(logger gokitlog.Logger, nodeName, bindAddr, bindPort, secret string, stopCh chan struct{}) (*SpeakerList, error) {
+	sl := SpeakerList{
+		l:      logger,
+		stopCh: stopCh,
+	}
+
+	if bindAddr == "" {
+		logger.Log("op", "startup", "msg", "Not starting fast dead node detection (MemberList), need ml-bindaddr config")
+		return &sl, nil
+	}
+
+	mconfig := memberlist.DefaultLANConfig()
+	// mconfig.Name MUST be spec.nodeName, as we will match it against Enpoints nodeName in usableNodes()
+	mconfig.Name = nodeName
+	mconfig.BindAddr = bindAddr
+	if bindPort != "" {
+		mlport, err := strconv.Atoi(bindPort)
+		if err != nil {
+			sl.l.Log("op", "startup", "error", "unable to parse ml-bindport", "msg", err)
+			return nil, err
+		}
+		mconfig.BindPort = mlport
+		mconfig.AdvertisePort = mlport
+	}
+	loggerout := gokitlog.NewStdlibAdapter(gokitlog.With(sl.l, "component", "MemberList"))
+	mconfig.Logger = golog.New(loggerout, "", golog.Lshortfile)
+	if secret != "" {
+		sha := sha256.New()
+		mconfig.SecretKey = sha.Sum([]byte(secret))[:16]
+	}
+	sl.eventCh = make(chan memberlist.NodeEvent, 16)
+	mconfig.Events = &memberlist.ChannelEventDelegate{Ch: sl.eventCh}
+	var err error
+	sl.mList, err = memberlist.Create(mconfig)
+	if err != nil {
+		sl.l.Log("op", "startup", "error", err, "msg", "failed to create memberlist")
+		return nil, err
+	}
+	return &sl, nil
+}
+
+// Start starts the needed goroutines
+func (sl *SpeakerList) Start(client *k8s.Client) {
+	sl.client = client
+
+	if sl.mList == nil {
+		return
+	}
+
+	go sl.watchEvents()
+	go sl.memberlistJoinLoop()
+}
+
+func (sl *SpeakerList) memberlistJoin() {
+	s := sl.k8sSpeakers.Load()
+	if s == nil {
+		return
+	}
+
+	mlSpeakers := map[string]bool{}
+	for _, n := range sl.mList.Members() {
+		mlSpeakers[n.Name] = true
+	}
+	k8sSpeakers := s.(map[string]string)
+	joinList := []string{}
+	for name, ip := range k8sSpeakers {
+		if _, ok := mlSpeakers[name]; !ok {
+			joinList = append(joinList, ip)
+		}
+	}
+	if len(joinList) > 0 {
+		n, err := sl.mList.Join(joinList)
+		sl.l.Log("op", "ml-join", "msg", "Memberlist join", "nb joigned", n, "error ?", err, "join list", joinList)
+	}
+}
+
+func (sl *SpeakerList) memberlistJoinLoop() {
+	ticker := time.NewTicker(30 * time.Second)
```

Comment:
This loop is only helpful during the node-monitor-grace-period (40s), after which disconnected pod is removed from endpoint list and rejoin is triggered by speakerlist update. Running at 30s internal doesn't help too much. I'm thinking such an optimization:

* Doesn't run memberlistJoinLoop in steady state
* trigger memberlistJoin on speakerlist and node event update to evaluate if there is a diff between speakerlist and member list
* If memberlistJoin notice a diff between speakerlist and member list, run the loop at a much smaller interval (5s?) until the diff has gone

Reply:
MemberList.Join takes 10s to timeout, and is sequential, so going to 5s might be too small. The current code is not optimal, maybe I could compute a "missing members" list on MemberList and k8sSpeakers changes, but I don't want the code to be too complicated either (we can't hold a mutex during join for example). I need to think about it more.

Reply:
good to know that. Even with that, I would suggest to keep the interval small when there is a diff. If Join takes longer than 5 sec, you just miss a tick and continue with next one. It's also likely to happen when you have 30 second interval. But good think about 5s is it makes it possible to quickly fix it when Join fails fast (due to icmp errors) or tcp timeout is configured small.

That can be left as an optimization.

Reply:
I went down to 10s